SeQuant/core/utility/singleton.hpp

```diff
@@ -75,21 +78,39 @@ class Singleton {
   /// default-constructible.
   /// @tparam Args a parameter pack type
   /// @param args a parameter pack
+  /// @return shared_ptr to the newly-created instance
   template <typename... Args>
-  static void set_instance(Args&&... args) {
-    assert(instance_accessor() == nullptr);
+  static std::shared_ptr<Derived> set_instance(Args&&... args) {
+    std::scoped_lock lock(instance_mutex());
     instance_accessor() = std::move(
-        std::unique_ptr<Derived>(new Derived(std::forward<Args>(args)...)));
+        std::shared_ptr<Derived>(new Derived(std::forward<Args>(args)...)));
```

Comment:
This should probably be
```
instance_accessor() = std::make_shared(std::forward<Args>(args)...);
```

Reply:
no, this is correct to allow using private ctors of Derived.